pkg/target/manager.go

```diff
@@ -0,0 +1,399 @@
+	ProjectClient(t Target) (client.Client, error)
+	SeedClient(t Target) (client.Client, error)
+	ShootNamespaceClient(t Target) (client.Client, error)
+	ShootClusterClient(t Target) (client.Client, error)
+}
+
+type managerImpl struct {
+	config          *Config
+	targetProvider  TargetProvider
+	clientProvider  ClientProvider
+	kubeconfigCache KubeconfigCache
+}
+
+var _ Manager = &managerImpl{}
+
+func NewManager(config *Config, targetProvider TargetProvider, clientProvider ClientProvider, kubeconfigCache KubeconfigCache) (Manager, error) {
+	return &managerImpl{
+		config:          config,
+		targetProvider:  targetProvider,
+		clientProvider:  clientProvider,
+		kubeconfigCache: kubeconfigCache,
+	}, nil
+}
+
+func (m *managerImpl) CurrentTarget() (Target, error) {
+	return m.targetProvider.Read()
+}
+
+func (m *managerImpl) TargetGarden(gardenName string) error {
+	for _, g := range m.config.Gardens {
+		if g.Name == gardenName {
+			return m.patchTarget(func(t *targetImpl) error {
+				t.Garden = gardenName
+				t.Project = ""
+				t.Seed = ""
+				t.Shoot = ""
+
+				return nil
+			})
+		}
+	}
+
+	return fmt.Errorf("garden %q is not defined in gardenctl configuration", gardenName)
+}
+
+func (m *managerImpl) TargetProject(ctx context.Context, projectName string) error {
+	return m.patchTarget(func(t *targetImpl) error {
+		if t.Garden == "" {
+			return ErrNoGardenTargeted
+		}
+
+		gardenClient, err := m.clientForGarden(t.Garden)
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		// validate that the project exists
+		if _, err := m.resolveProjectName(ctx, gardenClient, projectName); err != nil {
+			return fmt.Errorf("failed to validate project: %w", err)
+		}
+
+		t.Seed = ""
+		t.Project = projectName
+		t.Shoot = ""
+
+		return nil
+	})
+}
+
+func (m *managerImpl) resolveProjectName(ctx context.Context, gardenClient client.Client, projectName string) (*gardencorev1beta1.Project, error) {
+	// validate that the project exists
+	project := &gardencorev1beta1.Project{}
+	key := types.NamespacedName{Name: projectName}
+	err := gardenClient.Get(ctx, key, project)
+
+	return project, err
+}
+
+func (m *managerImpl) TargetSeed(ctx context.Context, seedName string) error {
+	return m.patchTarget(func(t *targetImpl) error {
+		if t.Garden == "" {
+			return ErrNoGardenTargeted
+		}
+
+		gardenClient, err := m.clientForGarden(t.Garden)
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		// validate that the seed exists
+		seed, err := m.resolveSeedName(ctx, gardenClient, seedName)
+		if err != nil {
+			return fmt.Errorf("failed to validate seed: %w", err)
+		}
+
+		t.Seed = seedName
+		t.Project = ""
+		t.Shoot = ""
+
+		if err := m.updateSeedKubeconfig(ctx, gardenClient, t, seed); err != nil {
+			return fmt.Errorf("failed to fetch kubeconfig for seed cluster: %w", err)
+		}
```

Comment:
this shouldn't be part of the patch func

Reply:
Why is that? If storing the kubeconfig locally fails, I do not want a "broken" target on disk.